File: generated_code/python/solution_1022_2.py
from typing import List

class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        # Time Complexity: O(4^(m * n)) in worst-case, but constrained heavily by the grid size (m * n <= 20)
        # Space Complexity: O(m * n) for the recursion stack
        # Algorithm: Backtracking (DFS) with state modifications to track visited cells.
        
        m, n = len(grid), len(grid[0])
        start_r, start_c = 0, 0
        empty_count = 0
        
        # Count the number of non-obstacle squares to visit and locate the starting position
        for r in range(m):
            for c in range(n):
                if grid[r][c] == 1:
                    start_r, start_c = r, c
                    empty_count += 1
                elif grid[r][c] == 0:
                    empty_count += 1
                elif grid[r][c] == 2:
                    empty_count += 1  # The target cell counts as a required step
                    
        paths_count = 0
        
        def backtrack(r: int, c: int, remain: int) -> None:
            nonlocal paths_count
            
            # Base case: if we reach the end square and have visited all required cells
            if grid[r][c] == 2:
                if remain == 1:
                    paths_count += 1
                return
            
            # Mark the current cell as visited using an obstacle placeholder (-1)
            temp = grid[r][c]
            grid[r][c] = -1
            
            # Explore all 4-directional neighbors
            for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                nr, nc = r + dr, c + dc
                if 0 <= nr < m and 0 <= nc < n and grid[nr][nc] != -1:
                    backtrack(nr, nc, remain - 1)
            
            # Undo the choice (backtrack) to restore the original grid state
            grid[r][c] = temp

        # Start the backtracking search from the starting square
        backtrack(start_r, start_c, empty_count)
        return paths_count

File: generated_code/cpp/solution_1022_2.cpp
#include <vector>

using namespace std;

class Solution {
private:
    int totalPaths = 0;
    int emptySquaresCount = 0;
    int targetRow = 0;
    int targetCol = 0;
    int rows = 0;
    int cols = 0;

    void backtrack(vector<vector<int>>& grid, int r, int c, int visitedCount) {
        // Base Case: If we reach the destination
        if (r == targetRow && c == targetCol) {
            // Check if we visited all non-obstacle squares (including the start square)
            if (visitedCount == emptySquaresCount) {
                totalPaths++;
            }
            return;
        }

        // 4-directional choices
        int dr[] = {-1, 1, 0, 0};
        int dc[] = {0, 0, -1, 1};

        for (int i = 0; i < 4; ++i) {
            int nr = r + dr[i];
            int nc = c + dc[i];

            // Validate boundaries and check if the square is walkable (0 or 2)
            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] >= 0) {
                int originalVal = grid[nr][nc];
                
                // Mark as visited using -2 to avoid revisit
                grid[nr][nc] = -2; 
                
                // Explore next step
                backtrack(grid, nr, nc, visitedCount + 1);
                
                // Backtrack and restore original value
                grid[nr][nc] = originalVal; 
            }
        }
    }

public:
    int uniquePathsIII(vector<vector<int>>& grid) {
        rows = grid.size();
        cols = grid[0].size();
        
        int startRow = 0, startCol = 0;
        emptySquaresCount = 0;
        totalPaths = 0;

        // Count required steps and locate start and end positions
        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {
                if (grid[r][c] != -1) {
                    emptySquaresCount++; // Count all valid squares (0, 1, 2)
                }
                if (grid[r][c] == 1) {
                    startRow = r;
                    startCol = c;
                } else if (grid[r][c] == 2) {
                    targetRow = r;
                    targetCol = c;
                }
            }
        }

        // Temporarily mark the starting position as visited
        grid[startRow][startCol] = -2;
        
        // Start backtracking from the starting position with initial count 1
        backtrack(grid, startRow, startCol, 1);

        return totalPaths;
    }
};